### DISCUSSION TITLE
post/android-productivity-and-more/

### DISCUSSION THREAD

**Discussion by giscus** [lyc8503/lyc8503#19] [Announcements]
# post/android-productivity-and-more/

本文主要记录我这两天关于如何将安卓手机/平板改造成生产力设备的一些探索和发现, 包含了软件/硬件相关的众多内容. 劝退首先先要说明, 研究这些方法的主要目的不是真的把安卓当电脑用或者是拿安卓当主力开发工具. 本文最终的结论也是, 最理想的情况下, 将安卓当作程序员的主力开发工具也会带来明显的生产力下降. 如果你确实有便携性的要求, 在此推荐购买体积 0.5L 左右迷你主机或者

https://blog.lyc8503.site/post/android-productivity-and-more/

**Comment by dabao1955** [comment]
感谢博主如此详细的教程~
补充我的一些思考以及勘误几点:
1.pkvm 方面，除了pixel 自家的 pkvm 以外，高通(sm8750 及以上，使用 gunyah)和联发科(d9400及以上，使用 GenizZone)也提供了类似的技术。且联发科可以运行未受保护的虚拟机。根据  gki_defconfig(第二点会提到)可以看到 CONFIG_KVM=y 是默认选项，但是这并不妨碍启动后会从 el2 降到 el1 导致 kvm 实际并不可用。
2.关于 gki 内核方面，实际上是从 5.10 开始(包括 5.10，因为引入了 abi 兼容这一个概念)才正式算作 gki 的范畴。5.4 应该算做 qgki(qualcomm gki，高通独有，和 google gki 不兼容)。
3.关于 android 运行 systemd ，除了 docker 和 lxc 一下，有一些小众项目则通过 cap 和 seccomp 来运行，例如 https://github.com/RuriOSS/ruri/blob/3746bbd27458e9d800d79b9f1fe726e9cb3277eb/doc/FAQ.md?plain=1#L13-L16
4.关于“可控的”可信”计算设备”，我觉得是不太可能的。一方面就是硬件层面的不可信。完全开源的硬件是不太可能的，市面上的开源硬件大多数都是 sdk 开源这种。另一方面，软件开源方面。lineage 的 vendor 就有很多非自由固件，就算不用 lineage，用 linux，linux对于手机的适配一方面是用爱发电另一方面就是适配性不太好这样。我觉得接近绝对可信的设备就两种，1是搭载 GrapheneOS 的 google pixel，另外一种就是像 pinephone 这种的 linux 手机。
以上
~~看在我打了这么多字的份上能给我 github 点个 follow 嘛~~

**Reply by lyc8503** [reply]
> 除了pixel 自家的 pkvm 以外，高通(sm8750 及以上，使用 gunyah)和联发科(d9400及以上，使用 GenizZone)也提供了类似的技术

对的对的，感谢补充，我创作本文时 8 Elite 和 9400 还没发布，后续的出厂安卓 >=16 的机器应该都需要支持 KVM 了；我后续其实也进一步研究过 KVM https://blog.lyc8503.net/post/android-kvm-on-mediatek/

> 关于 gki 内核方面，实际上是从 5.10 开始(包括 5.10，因为引入了 abi 兼容这一个概念)才正式算作 gki 的范畴。5.4 应该算做 qgki(qualcomm gki，高通独有，和 google gki 不兼容)。

确实，好像写错了，当时可能查到了不精准的信息，感谢指正

> 关于“可控的”可信”计算设备”，我觉得是不太可能的

哦，这点我好像用词不恰当引发了一些误解，我文中开头的"可信"并非指可信计算那种程度的可信，只是想强调设备本身是自己所有的。不过看现在的这个情况，可能过两年真的只有 Pixel 甚至 Linux 手机可以用了。

> ~~看在我打了这么多字的份上能给我 github 点个 follow 嘛~~

~~followed，您正好是我的第 100 个 following 哎~~

**Reply by dabao1955** [reply]
> 我后续其实也进一步研究过 KVM https://blog.lyc8503.net/post/android-kvm-on-mediatek/

其实 pkvm 和 kvm 还是不太一样的，~~pixel 和 mediatek 可以绕过限制运行不受保护的虚拟机~~。首先 pkvm 一般不会暴露 kvm 接口，其次 pkvm 有专门的工具 crosvm 来通过 pkvm 运行虚拟机
